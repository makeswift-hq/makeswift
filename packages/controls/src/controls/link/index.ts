export * from './link'
export * from './generic-link'
