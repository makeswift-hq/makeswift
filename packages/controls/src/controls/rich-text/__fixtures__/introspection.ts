import { RichTextDTO } from '../dto'

export const SWATCH_ID = 'SWATCH_ID='
export const TYPOGRAPHY_ID = 'TYPOGRAPHY_ID='
export const PAGE_ID = 'PAGE_ID='

export const introspection: RichTextDTO = {
  object: 'value',
  annotations: undefined,
  data: undefined,
  document: {
    data: undefined,
    nodes: [
      {
        data: {},
        type: 'blockquote',
        nodes: [
          {
            object: 'inline',
            type: 'link',
            data: {
              type: 'OPEN_PAGE',
              payload: {
                pageId: PAGE_ID,
                openInNewTab: false,
              },
            },
            nodes: [
              {
                text: 'Credibly implement global synergy, then collaboratively implement goal-oriented sprints. Rapidiously initiate standards compliant clouds. ',
                marks: [
                  {
                    data: {
                      value: {
                        id: TYPOGRAPHY_ID,
                        style: [
                          {
                            value: {
                              color: {
                                swatchId: SWATCH_ID,
                                alpha: 1,
                              },
                            },
                          },
                          {
                            value: {
                              fontWeight: 500,
                            },
                          },
                          {
                            value: {
                              color: {
                                swatchId: undefined,
                                alpha: 1,
                              },
                            },
                          },
                        ],
                      },
                    },
                    type: 'typography',
                    object: 'mark',
                  },
                ],
                object: 'text',
              },
            ],
          },
        ],
        object: 'block',
      },
    ],
    object: 'document',
  },
  selection: {
    focus: {
      path: [1, 2],
      object: 'point',
      offset: 0,
    },
    marks: undefined,
    anchor: {
      path: [0, 0],
      object: 'point',
      offset: 0,
    },
    object: 'selection',
    isFocused: false,
  },
}
