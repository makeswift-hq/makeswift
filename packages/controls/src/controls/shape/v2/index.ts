export * from './shape-v2'
