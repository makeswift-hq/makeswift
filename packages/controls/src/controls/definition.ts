import { z } from 'zod'

import { type ParseResult } from '../lib/zod'

import { type Data } from '../common/types'
import {
  type CopyContext,
  type MergeContext,
  type MergeTranslatableDataContext,
} from '../context'
import { type IntrospectionTarget } from '../introspection'
import {
  Serializable,
  serializeObject,
  type SerializedRecord,
} from '../serialization'

import { ControlInstance, type SendMessage } from './instance'

export type SchemaType<T> = z.ZodType<T>
export type SchemaTypeAny = SchemaType<any> | z.ZodBranded<SchemaType<any>, any>

export type Resolvable<T> = {
  readStableValue(previous?: T): T
  subscribe(onUpdate: () => void): () => void
}

// workaround for TS type inference issues
export type UndefinedOr<T> = T | undefined

export abstract class ControlDefinition<
  ControlType extends string = string,
  Config = unknown,
  DataType extends Data = Data,
  ValueType extends Data = Data,
  ResolvedValueType = Data | unknown,
  InstanceType extends ControlInstance<any> = ControlInstance<any>,
> extends Serializable {
  constructor(readonly config: Config) {
    super()
  }

  abstract get controlType(): ControlType

  abstract get schema(): {
    definition: SchemaType<unknown>
    type: SchemaType<ControlType>
    data: SchemaType<DataType>
    value: SchemaType<ValueType>
    resolvedValue: SchemaType<ResolvedValueType>
  }

  abstract safeParse(
    data: unknown | undefined,
  ): ParseResult<UndefinedOr<DataType>>

  abstract fromData(data: UndefinedOr<DataType>): UndefinedOr<ValueType>
  abstract toData(value: ValueType): DataType

  abstract copyData(
    data: UndefinedOr<DataType>,
    context: CopyContext,
  ): UndefinedOr<DataType>

  mergeData(
    base: DataType,
    override: UndefinedOr<DataType>,
    _context: MergeContext,
  ): DataType {
    return override ?? base
  }

  getTranslatableData(_data: DataType): Data {
    return null
  }

  mergeTranslatedData(
    data: DataType,
    _translatedData: Data,
    _context: MergeTranslatableDataContext,
  ): Data {
    return data as Data
  }

  resolveValue(
    _data: UndefinedOr<DataType>,
  ): Resolvable<UndefinedOr<ResolvedValueType>> {
    console.assert(
      false,
      `${this.controlType}: 'resolveValue' is not implemented`,
    )

    return {
      readStableValue: (_previous?: ResolvedValueType) => undefined,
      subscribe: () => () => {},
    }
  }

  abstract createInstance(sendMessage: SendMessage<any>): InstanceType

  abstract serialize(): [SerializedRecord, Transferable[]]

  introspect<R>(
    data: UndefinedOr<DataType>,
    target: IntrospectionTarget<R>,
  ): R[] {
    return target.introspect(data)
  }
}

export function serialize(
  config: unknown,
  rest: { type: string } & Record<string, unknown>,
): [SerializedRecord, Transferable[]] {
  const [serializedConfig, transferables] = serializeObject(config)
  return [
    {
      config: serializedConfig,
      ...rest,
    } as unknown as SerializedRecord,
    transferables,
  ]
}
