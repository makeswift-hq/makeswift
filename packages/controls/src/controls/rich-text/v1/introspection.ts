import { type IntrospectionTarget } from '../../../introspection'
import { type DataType } from '../../associated-types'
import { GenericLink, LinkDefinition } from '../../link'
import { unstable_Typography } from '../../typography'

import { type RichTextDTO } from '../dto'
import {
  TextJSON,
  type InlineJSON,
  type MarkJSON,
  type NodeJSON,
} from '../dto/types'

export function introspectRichTextData<R>(
  data: RichTextDTO | undefined,
  target: IntrospectionTarget<R>,
): R[] {
  if (data == null || data.document == null) return []

  function introspectNode(node: NodeJSON): R[] {
    switch (node.object) {
      case 'document':
      case 'block':
        return node.nodes?.flatMap(introspectNode) ?? []

      case 'inline':
      case 'text':
        return introspectInlineOrTextNode(node)

      default:
        return []
    }
  }

  function introspectInlineOrTextNode(node: InlineJSON | TextJSON): R[] {
    switch (node.object) {
      case 'inline':
        return [
          ...introspectInlineNode(node),
          ...(node.nodes?.flatMap(introspectInlineOrTextNode) ?? []),
        ]

      case 'text':
        return node.marks?.flatMap(introspectMark) ?? []

      default:
        return []
    }
  }

  function introspectInlineNode(inline: InlineJSON): R[] {
    switch (inline.type) {
      case 'link': {
        return GenericLink().introspect(
          inline.data as DataType<LinkDefinition>,
          target,
        )
      }

      default:
        return []
    }
  }

  function introspectMark(mark: MarkJSON): R[] {
    const typographyResult = unstable_Typography().introspect(
      mark.data?.value,
      target,
    )

    return [...typographyResult]
  }

  return introspectNode(data.document)
}
