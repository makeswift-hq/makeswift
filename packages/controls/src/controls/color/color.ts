import { match } from 'ts-pattern'
import { z } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import { ControlDataTypeKey, Schema } from '../../common'
import { type CopyContext } from '../../context'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import { ControlDefinition, serialize, type SchemaType } from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'

type Config = z.infer<typeof Definition.schema.relaxed.config>

type SchemaByDefaultValue<D extends Config['defaultValue']> =
  undefined extends D
    ? typeof Definition.schema.relaxed
    : typeof Definition.schema.strict

type Schema<C extends Config> = SchemaByDefaultValue<C['defaultValue']>
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<C extends Config> = z.infer<Schema<C>['resolvedValue']>

type ReturnedSchemaType<C extends Config> = {
  definition: typeof Definition.schema.relaxed.definition
  type: typeof Definition.schema.relaxed.type
  data: SchemaType<DataType<C>>
  value: SchemaType<ValueType<C>>
  resolvedValue: SchemaType<ResolvedValueType<C>>
}

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>
> {
  private static readonly v1DataType = 'color::v1' as const
  private static readonly dataSignature = {
    v1: { [ControlDataTypeKey]: this.v1DataType },
  } as const

  static readonly type = 'makeswift::controls::color' as const

  static get schema() {
    const type = z.literal(this.type)
    const version = z.literal(1).optional()
    const value = Schema.colorData
    const resolvedSwatchValue = Schema.resolvedColorData

    const schemas = <R>(resolvedValue: SchemaType<R>) => {
      const data = value.merge(
        z.object({
          [ControlDataTypeKey]: z.literal(this.v1DataType).optional(),
        }),
      )

      const config = z.object({
        label: z.string().optional(),
        labelOrientation: z
          .union([z.literal('horizontal'), z.literal('vertical')])
          .optional(),
        defaultValue: resolvedValue,
        hideAlpha: z.boolean().optional(),
      })

      const definition = z.object({
        type,
        config,
        version,
      })

      return {
        type,
        value,
        resolvedSwatchValue,
        resolvedValue,
        data,
        config,
        version,
        definition,
      }
    }

    return {
      version,
      relaxed: schemas(z.string().optional()),
      strict: schemas(z.string()),
    }
  }

  static deserialize(data: DeserializedRecord): ColorDefinition {
    if (data.type !== Definition.type)
      throw new Error(
        `Color deserialization: expected '${Definition.type}', got '${data.type}'`,
      )

    const { config, version } = Definition.schema.relaxed.definition.parse(data)
    return new ColorDefinition(config, version)
  }

  constructor(
    config: C,
    readonly version: z.infer<typeof Definition.schema.version>,
  ) {
    super(config)
  }

  get controlType() {
    return Definition.type
  }

  get schema(): ReturnedSchemaType<C> {
    return Definition.schema.relaxed
  }

  get dataSchema() {
    return (
      (this.config.defaultValue === undefined
        ? Definition.schema.relaxed
        : Definition.schema.strict) as Schema<C>
    ).data
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.dataSchema, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, ({ swatchId, alpha }) => ({
        swatchId,
        alpha,
      }))
      .otherwise((value) => value)
  }

  toData(value: ValueType<C>): DataType<C> {
    return match(this.version)
      .with(1, () => ({
        ...Definition.dataSignature.v1,
        ...value,
      }))
      .with(undefined, () => value)
      .exhaustive()
  }

  copyData(
    data: DataType<C> | undefined,
    { replacementContext }: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return data

    const replaceSwatchId = (swatchId: string) =>
      replacementContext.swatchIds.get(swatchId) ?? swatchId

    const inputSchema = this.dataSchema.optional()
    return match(data satisfies z.infer<typeof inputSchema>)
      .with(Definition.dataSignature.v1, (val) => ({
        ...val,
        swatchId: replaceSwatchId(val.swatchId),
      }))
      .otherwise((val) => ({
        ...val,
        swatchId: replaceSwatchId(val.swatchId),
      }))
  }

  createInstance(sendMessage: SendMessage<any>) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
      version: this.version,
    })
  }
}

export class ColorDefinition<C extends Config = Config> extends Definition<C> {}

type UserConfig<D extends Config['defaultValue']> = Config & {
  defaultValue?: D
}

type NormedConfig<D extends Config['defaultValue']> = z.infer<
  SchemaByDefaultValue<D>['config']
>

export function Color<D extends Config['defaultValue']>(
  config?: UserConfig<D>,
): ColorDefinition<NormedConfig<D>> {
  return new ColorDefinition<NormedConfig<D>>(config ?? {}, 1)
}
