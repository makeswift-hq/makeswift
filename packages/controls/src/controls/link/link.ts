import { z } from 'zod'

import { safeParse, type ParseResult } from '../../lib/zod'

import { type CopyContext } from '../../context'
import { Targets, type IntrospectionTarget } from '../../introspection'
import { type SerializedRecord } from '../../serialization'

import { ControlDefinition, serialize, type SchemaType } from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'

import { type GenericMouseEvent } from './mouse-event'
import { linkSchema } from './schema'

type Config = z.infer<typeof Definition.schema.config>

type Schema<_C extends Config> = typeof Definition.schema
type LinkTarget<C extends Config> = z.infer<Schema<C>['linkTarget']>
type DataType<C extends Config> = z.infer<Schema<C>['data']>
type ValueType<C extends Config> = z.infer<Schema<C>['value']>
type ResolvedValueType<
  MouseEventType extends GenericMouseEvent,
  C extends Config,
> = {
  href: string
  target?: LinkTarget<C>
  onClick: (event: MouseEventType) => void
}

class Definition<
  MouseEventType extends GenericMouseEvent = GenericMouseEvent,
  C extends Config = Config,
> extends ControlDefinition<
  typeof Definition.type,
  Config,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<MouseEventType, C>
> {
  static readonly type = 'makeswift::controls::link' as const

  static get schema() {
    const type = z.literal(this.type)

    const link = linkSchema
    const data = z.union([link, z.null()])
    const value = data

    const config = z.object({
      label: z.string().optional(),
    })

    const definition = z.object({
      type,
      config,
    })

    const linkTarget = z.enum(['_blank', '_self'])

    const resolvedValue = z.object({
      href: z.string(),
      target: linkTarget.optional(),
      onClick: z.function().args(z.any()).returns(z.void()),
    })

    return {
      type,
      data,
      link,
      linkTarget,
      value,
      resolvedValue,
      config,
      definition,
    }
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    return {
      ...Definition.schema,
      data: Definition.schema.data as SchemaType<DataType<C>>,
      value: Definition.schema.value as SchemaType<ValueType<C>>,
      resolvedValue: Definition.schema.resolvedValue as SchemaType<
        ResolvedValueType<MouseEventType, C>
      >,
    }
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    return this.schema.data.optional().parse(data)
  }

  toData(value: ValueType<C>): DataType<C> {
    return value
  }

  copyData(
    data: DataType<C> | undefined,
    { replacementContext }: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return data

    if (data.type === 'OPEN_PAGE') {
      const { pageId } = data.payload
      if (pageId != null) {
        return {
          ...data,
          payload: {
            ...data.payload,
            pageId: replacementContext.pageIds.get(pageId) ?? pageId,
          },
        }
      }
    }

    if (data.type === 'SCROLL_TO_ELEMENT') {
      const { elementIdConfig } = data.payload
      if (elementIdConfig == null) return data

      const { elementKey } = elementIdConfig
      return {
        ...data,
        payload: {
          ...data.payload,
          elementIdConfig: {
            ...elementIdConfig,
            elementKey:
              replacementContext.elementKeys.get(elementKey) ?? elementKey,
          },
        },
      }
    }

    return data
  }

  introspect<R>(
    data: DataType<C> | undefined,
    target: IntrospectionTarget<R>,
  ): R[] {
    if (target.type !== Targets.Page.type) return []
    if (data == null) return []

    switch (data.type) {
      case 'OPEN_PAGE':
        return (data.payload.pageId != null ? [data.payload.pageId] : []) as R[]

      default:
        return []
    }
  }

  createInstance(sendMessage: SendMessage<any>) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }
}

export { Definition as LinkDefinition, type GenericMouseEvent }
