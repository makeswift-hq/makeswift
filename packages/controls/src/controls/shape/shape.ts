import { z } from 'zod'

import { mapValues } from '../../lib/functional'
import { safeParse, type ParseResult } from '../../lib/zod'

import { type Data } from '../../common'
import {
  type CopyContext,
  type MergeTranslatableDataContext,
} from '../../context'
import { type IntrospectionTarget } from '../../introspection'
import {
  SerializationSchema,
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'

import {
  type DataType as DataType_,
  type ResolvedValueType as ResolvedValueType_,
  type ValueType as ValueType_,
} from '../associated-types'
import { ControlDefinition, serialize, type SchemaType } from '../definition'
import { type SendMessage } from '../instance'

import { ShapeControl } from './shape-control'

type ItemDefinition = ControlDefinition<string, unknown, any, any, any>
type KeyDefinitions = Record<string, ItemDefinition>

type Config<Defs extends KeyDefinitions = KeyDefinitions> = {
  readonly type: Defs
}

type DataType<C extends Config> = {
  [K in keyof C['type']]?: DataType_<C['type'][K]>
}

type ValueType<C extends Config> = {
  [K in keyof C['type']]?: ValueType_<C['type'][K]>
}

type ResolvedValueType<C extends Config> = {
  [K in keyof C['type']]: ResolvedValueType_<C['type'][K]>
}

type InstanceType<C extends Config> = ShapeControl<Definition<C>>

class Definition<C extends Config> extends ControlDefinition<
  typeof Definition.type,
  C,
  DataType<C>,
  ValueType<C>,
  ResolvedValueType<C>,
  InstanceType<C>
> {
  static readonly type = 'makeswift::controls::shape' as const

  static deserialize(
    data: DeserializedRecord,
    deserializeCallback: (r: DeserializedRecord) => ControlDefinition,
  ): ShapeDefinition {
    if (data.type !== Definition.type)
      throw new Error(
        `Shape: expected '${Definition.type}', got '${data.type}'`,
      )

    const {
      config: { type, ...config },
    } = z
      .object({
        type: z.literal(Definition.type),
        config: z.object({
          type: z.record(z.string(), SerializationSchema.deserializedRecord),
        }),
      })
      .parse(data)

    const deserializedType = mapValues(type, (itemDef) => {
      return deserializeCallback(itemDef)
    })

    return Shape({
      type: deserializedType,
      ...config,
    })
  }

  get keyDefs() {
    return this.config.type
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    const type = z.literal(Definition.type)
    const keys = mapValues(this.keyDefs, (def) => def.schema)

    const data = z.object(mapValues(keys, (key) => key.data)) as SchemaType<
      DataType<C>
    >

    const value = z.object(mapValues(keys, (key) => key.value)) as SchemaType<
      ValueType<C>
    >

    const resolvedValue = z.object(
      mapValues(keys, (key) => key.resolvedValue),
    ) as SchemaType<ResolvedValueType<C>>

    const config = z.object({
      type: z.object(mapValues(keys, (key) => key.definition)),
    })

    const definition = z.object({
      type,
      config,
    })

    return { type, data, value, resolvedValue, definition }
  }

  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
    if (data == null) return undefined
    return mapValues(data, (value, key) =>
      this.keyDefs[key as string]?.fromData(value),
    )
  }

  toData(value: ValueType<C>): DataType<C> {
    return mapValues(value, (value, key) =>
      this.keyDefs[key as string]?.toData(value),
    )
  }

  copyData(
    data: DataType<C> | undefined,
    context: CopyContext,
  ): DataType<C> | undefined {
    if (data == null) return undefined
    return mapValues(data, (value, key) =>
      this.keyDefs[key as string].copyData(value, context),
    )
  }

  getTranslatableData(data: DataType<C>): Data {
    if (data == null) return null
    return mapValues(data, (value, key) => {
      return this.keyDefs[key as string]?.getTranslatableData(value)
    })
  }

  mergeTranslatedData(
    data: DataType<C>,
    translatedData: Record<string, DataType<C>>,
    context: MergeTranslatableDataContext,
  ): Data {
    if (translatedData == null) return data
    return mapValues(data, (value, key) =>
      this.keyDefs[key as string]?.mergeTranslatedData(
        value,
        translatedData[key as string],
        context,
      ),
    )
  }

  createInstance(sendMessage: SendMessage): InstanceType<C> {
    return new ShapeControl(this, sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  introspect<R>(data: DataType<C> | undefined, target: IntrospectionTarget<R>) {
    return Object.entries(data ?? {}).flatMap(([key, value]) =>
      this.keyDefs[key as string]?.introspect(value, target),
    )
  }
}

export class ShapeDefinition<C extends Config = Config> extends Definition<C> {}

export function Shape<Defs extends KeyDefinitions>(
  config: Config<Defs>,
): ShapeDefinition<Config<Defs>> {
  return new ShapeDefinition<Config<Defs>>(config)
}
