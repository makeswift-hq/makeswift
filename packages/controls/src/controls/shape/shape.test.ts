import { Targets } from '../../introspection'
import { deserializeRecord, type DeserializedRecord } from '../../serialization'

import { Checkbox, CheckboxDefinition } from '../checkbox'
import { Color, ColorDefinition } from '../color'
import { Combobox } from '../combobox'
import { ControlDefinition } from '../definition'
import { Image } from '../image'
import { GenericLink as Link } from '../link'
import { TextArea } from '../text-area'
import { TextInput } from '../text-input'

import { Shape, ShapeDefinition } from './shape'

function deserializer(record: DeserializedRecord): ControlDefinition {
  switch (record.type) {
    case ShapeDefinition.type:
      return ShapeDefinition.deserialize(record, deserializer)

    case ColorDefinition.type:
      return ColorDefinition.deserialize(record)

    case CheckboxDefinition.type:
      return CheckboxDefinition.deserialize(record)

    default:
      throw new Error('Invalid control type')
  }
}

describe('Shape', () => {
  describe('constructor', () => {
    test('heterogenous controls', () => {
      const shape = Shape({
        type: {
          val: Color({ defaultValue: 'red' }),
          num: Checkbox(),
        },
      })
      expect(shape).toMatchSnapshot()
    })

    test('disallows extraneous properties', () => {
      Shape({
        type: {
          toggle: Checkbox(),
        },
        // @ts-expect-error
        extra: 'extra',
      })
    })
  })

  describe('assignability', () => {
    function assignTest(_def: ShapeDefinition) {}
    assignTest(
      Shape({
        type: {},
      }),
    )

    assignTest(
      Shape({
        type: {
          toggle: Checkbox(),
          color: Color({ defaultValue: 'red' }),
        },
      }),
    )

    assignTest(
      Shape({
        type: {
          combo: Combobox({
            getOptions: () => [{ id: '0', value: 'a', label: 'Alpha' }],
          }),
          link: Link({ label: 'Link' }),
        },
      }),
    )
  })

  describe('serialization', () => {
    test('serialize/deserialize shape of controls', () => {
      const shape = Shape({
        type: {
          color: Color({ defaultValue: 'red' }),
          checkbox: Checkbox(),
        },
      })

      const [serialized, _] = shape.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = ShapeDefinition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(shape)
    })

    test('serialize/deserialize composable shape of controls', () => {
      const shape = Shape({
        type: {
          shape: Shape({
            type: {
              color: Color({ defaultValue: 'red' }),
              checkbox: Checkbox(),
            },
          }),
        },
      })

      const [serialized, _] = shape.serialize()
      expect(serialized).toMatchSnapshot()

      const deserialized = ShapeDefinition.deserialize(
        deserializeRecord(serialized),
        deserializer,
      )

      expect(deserialized).toEqual(shape)
    })
  })

  describe('introspection', () => {
    test('is correctly delegated to subcontrols', () => {
      const shape = Shape({
        type: {
          color: Color({ defaultValue: 'red' }),
          link: Link(),
          image: Image(),
        },
      })

      const shapeData = {
        color: {
          swatchId: 'swatch-id',
          alpha: 1,
        },
        link: {
          type: 'OPEN_PAGE' as const,
          payload: {
            pageId: 'page-id',
            openInNewTab: false,
          },
        },
        image: 'file-id',
      }

      const swatchIds = shape.introspect(shapeData, Targets.Swatch)
      expect(swatchIds).toEqual(['swatch-id'])

      const pageIds = shape.introspect(shapeData, Targets.Page)
      expect(pageIds).toEqual(['page-id'])

      const fileIds = shape.introspect(shapeData, Targets.File)
      expect(fileIds).toEqual(['file-id'])
    })
  })

  describe('getTranslatableData', () => {
    test('is correctly delegated to subcontrols', () => {
      const shape = Shape({
        type: {
          textArea: TextArea(),
          textInput: TextInput(),
          color: Color({ defaultValue: 'red' }),
        },
      })

      const shapeData = {
        textArea: 'alpha',
        textInput: 'beta',
        color: {
          swatchId: 'swatch-id',
          alpha: 1,
        },
      }

      const translatableData = shape.getTranslatableData(shapeData)
      expect(translatableData).toEqual({
        textArea: 'alpha',
        textInput: 'beta',
        color: null,
      })
    })
  })
})
