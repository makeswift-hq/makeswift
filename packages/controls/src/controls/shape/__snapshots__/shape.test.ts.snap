// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shape constructor heterogenous controls 1`] = `
ShapeDefinition {
  "config": {
    "type": {
      "num": CheckboxDefinition {
        "config": {},
        "version": 1,
      },
      "val": ColorDefinition {
        "config": {
          "defaultValue": "red",
        },
        "version": 1,
      },
    },
  },
}
`;

exports[`Shape serialization serialize/deserialize composable shape of controls 1`] = `
{
  "config": {
    "type": {
      "shape": {
        "config": {
          "type": {
            "checkbox": {
              "config": {},
              "type": "makeswift::controls::checkbox",
              "version": 1,
            },
            "color": {
              "config": {
                "defaultValue": "red",
              },
              "type": "makeswift::controls::color",
              "version": 1,
            },
          },
        },
        "type": "makeswift::controls::shape",
      },
    },
  },
  "type": "makeswift::controls::shape",
}
`;

exports[`Shape serialization serialize/deserialize shape of controls 1`] = `
{
  "config": {
    "type": {
      "checkbox": {
        "config": {},
        "type": "makeswift::controls::checkbox",
        "version": 1,
      },
      "color": {
        "config": {
          "defaultValue": "red",
        },
        "type": "makeswift::controls::color",
        "version": 1,
      },
    },
  },
  "type": "makeswift::controls::shape",
}
`;
